### PR TITLE
feat: optional password validation in updateEmailOrPassword

### PR DESCRIPTION
**Pull request by iresharma**
## Summary of change
Added optional password validation in `updateEmailOrPassword` function.

## Related issues

-   https://github.com/supertokens/supertokens-node/issues/549 (same implemented for golang)

## Documentation changes

(If relevant, please create a PR in our [docs repo](https://github.com/supertokens/docs), or create a checklist here highlighting the necessary changes)

## Checklist for important updates

-   [x] Changelog has been updated
-   [ ] `coreDriverInterfaceSupported.json` file has been updated (if needed)
    -   Along with the associated array in `supertokens/constants.go`
-   [ ] `frontendDriverInterfaceSupported.json` file has been updated (if needed)
-   [x] Changes to the version if needed
    -   In `supertokens/constants.go > version variable`
-   [x] Had installed and ran the pre-commit hook
-   [ ] Issue this PR against the latest non released version branch.
    -   To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
    -   If no such branch exists, then create one from the latest released branch.

## Remaining TODOs for this PR

-   [ ] Item1
-   [ ] Item2

### CODE REVIEW

File: recipe/emailpassword/recipeImplementation.go [hunk 1]
```diff
@@ -158,14 +158,28 @@ func MakeRecipeImplementation(querier supertokens.Querier) epmodels.RecipeInterf
...
 		requestBody := map[string]interface{}{
 			"userId": userId,
 		}
 		if email != nil {
 			requestBody["email"] = email
 		}
 		if password != nil {
+			if applyPasswordPolicy == nil || *applyPasswordPolicy {
+				formFields := getEmailPasswordConfig().SignUpFeature.FormFields
+				for i := range formFields {
```
Comment (iresharma) [inline recipe/emailpassword/recipeImplementation.go:171]:
```suggestion
				for i := range formFields {
					if formFields[i].ID == "password" {
						err := formFields[i].Validate(*password)
						if err != nil {
							errResponse := epmodels.PasswordPolicyViolatedError{
								FailureReason: err,
							}
							return epmodels.UpdateEmailOrPasswordResponse{PasswordPolicyViolatedError: &errResponse}, nil
						}
					}
					break;
				}
```

Should I add a break statement here, there can be a lot of fields but as soon as password field is validated we can break

File: recipe/emailpassword/epmodels/recipeInterface.go [hunk 2]
```diff
@@ -56,7 +56,12 @@ type ResetPasswordUsingTokenResponse struct {
...
-	UnknownUserIdError      *struct{}
-	EmailAlreadyExistsError *struct{}
+	OK                          *struct{}
+	UnknownUserIdError          *struct{}
+	EmailAlreadyExistsError     *struct{}
+	PasswordPolicyViolatedError *PasswordPolicyViolatedError
+}
+
+type PasswordPolicyViolatedError struct {
+	FailureReason *string
```
Comment (rishabhpoddar) [inline recipe/emailpassword/epmodels/recipeInterface.go:66]:
```suggestion
	FailureReason string
```

Comment (rishabhpoddar) [reply recipe/emailpassword/epmodels/recipeInterface.go:66]: Cause if this struct is defined, there will always be a reason.

File: recipe/thirdpartyemailpassword/main.go [hunk 3]
```diff
@@ -91,12 +91,12 @@ func ResetPasswordUsingTokenWithContext(token, newPassword string, userContext s
 	return (*instance.RecipeImpl.ResetPasswordUsingToken)(token, newPassword, userContext)
 }
 
-func UpdateEmailOrPasswordWithContext(userId string, email *string, password *string, userContext supertokens.UserContext) (epmodels.UpdateEmailOrPasswordResponse, error) {
+func UpdateEmailOrPasswordWithContext(userId string, email *string, password *string, applyPasswordPOlicy *bool, userContext supertokens.UserContext) (epmodels.UpdateEmailOrPasswordResponse, error) {
```
[the remark this answers is not in the document]
Comment (rishabhpoddar) [reply recipe/thirdpartyemailpassword/main.go:94]: Capital "O" in POlicy.. Please double check your PRs.

File: recipe/thirdpartyemailpassword/main.go [hunk 4]
```diff
@@ -91,12 +91,12 @@ func ResetPasswordUsingTokenWithContext(token, newPassword string, userContext s
...
 	instance, err := GetRecipeInstanceOrThrowError()
 	if err != nil {
 		return epmodels.UpdateEmailOrPasswordResponse{}, err
 	}
-	return (*instance.RecipeImpl.UpdateEmailOrPassword)(userId, email, password, userContext)
+	return (*instance.RecipeImpl.UpdateEmailOrPassword)(userId, email, password, applyPasswordPOlicy, userContext)
```
Comment (rishabhpoddar) [inline recipe/thirdpartyemailpassword/main.go:99]:
```suggestion
	return (*instance.RecipeImpl.UpdateEmailOrPassword)(userId, email, password, applyPasswordPolicy, userContext)
```

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: recipe/emailpassword/updateEmailPass_test.go [hunk 5]
```diff
@@ -140,6 +139,12 @@ func TestUpdateEmailPass(t *testing.T) {
 
 	assert.Equal(t, "OK", data2["status"])
 	assert.Equal(t, email, data2["user"].(map[string]interface{})["email"])
+
+	password = "test"
+	applyPasswordPolicy := true
+	res3, err := UpdateEmailOrPassword(data["user"].(map[string]interface{})["id"].(string), &email, &password, &applyPasswordPolicy)
+	assert.NotNil(t, res3.PasswordPolicyViolatedError)
+	assert.Equal(t, "Password must contain at least 8 characters, including a number", res3.PasswordPolicyViolatedError.FailureReason)
 }
```
Comment (rishabhpoddar) [inline recipe/emailpassword/updateEmailPass_test.go:148]: add more tests

File: recipe/thirdpartyemailpassword/recipe.go [hunk 6]
```diff
@@ -54,6 +54,7 @@ func MakeRecipe(recipeId string, appInfo supertokens.NormalisedAppinfo, config *
 		return Recipe{}, err
 	}
 	r.Config = verifiedConfig
+	var emailPasswordRecipe emailpassword.Recipe
```
Comment (rishabhpoddar) [inline recipe/thirdpartyemailpassword/recipe.go:57]: undo change for where this variable was

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

**Review by rishabhpoddar — APPROVED** [review]

**Pull request merged**